onnx/defs/math/defs.cc

```diff
@@ -3337,4 +3337,220 @@ ONNX_OPERATOR_SET_SCHEMA(
           result_shape_proto.add_dim()->set_dim_value(2);
           updateOutputShape(ctx, 0, result_shape_proto);
         }));
+
+static const char* SVD_doc = R"DOC(
+Singular Value Decomposition
+
+Factorizes A using SVD. Both full and partial SVD compute the standard SVD factorization `A = U * diag(S) * Vh`.
+Computes the full SVD by default. Partial SVD is also known as [thin SVD](https://en.wikipedia.org/wiki/Singular_value_decomposition#Thin_SVD)
+
+A must have at least 2 dimensions.
+
+All dimensions higher than 2 are used for broadcasting, and SVD is only applied to the lowest two dimensions.
+See [numpy](https://numpy.org/doc/stable/reference/generated/numpy.linalg.svd.html).
+
+Dimensions:
+
+A: (M, N)
+
+K: Min(M, N)
+
+Full SVD:
+
+U: (M, M)
+
+S: (K)
+
+Vh: (N, N)
+
+Partial SVD:
+
+U: (M, K)
+
+S: (K)
+
+Vh: (K, N)
+
+TODO - add doc about differentiability
+)DOC";
+
+// When compute_uv is set to false
+// - inference is not performed for U and Vh
+// - S is at output index 0
+void SVDInferenceFunction(InferenceContext& ctx, bool compute_uv) {
+  const auto U_idx = 0;
+  const auto S_idx = compute_uv ? 1 : 0;
+  const auto Vh_idx = 2;
+
+  // All output elements types are same as input element types
+  if (compute_uv) {
+    propagateElemTypeFromInputToOutput(ctx, 0, U_idx);
+    propagateElemTypeFromInputToOutput(ctx, 0, Vh_idx);
+  }
+
+  propagateElemTypeFromInputToOutput(ctx, 0, S_idx);
+
+  // SVD is applied to lowest two dimensions of the matrix.
+  // The rest of the dimensions are used for broadcasting,
+  // see https://numpy.org/doc/stable/reference/generated/numpy.linalg.svd.html
+
+  const auto A_shape = ctx.getInputType(0)->tensor_type().shape();
+  const auto A_dim_size = A_shape.dim_size();
+
+  if (A_dim_size < 2) {
+    fail_shape_inference("Input tensor must have at least 2 dimensions");
+  }
+
+  auto dim_idx = 0;
+
+  auto U_shape = compute_uv ? ctx.getOutputType(U_idx)->mutable_tensor_type()->mutable_shape() : nullptr;
+  auto S_shape = ctx.getOutputType(S_idx)->mutable_tensor_type()->mutable_shape();
+  auto Vh_shape = compute_uv ? ctx.getOutputType(Vh_idx)->mutable_tensor_type()->mutable_shape() : nullptr;
+
+  // Copy over all dimensions but the last two
+  for (; dim_idx < A_dim_size - 2; ++dim_idx) {
+    const auto dim = A_shape.dim(dim_idx);
+
+    if (compute_uv) {
+      *U_shape->add_dim() = dim;
+      *Vh_shape->add_dim() = dim;
+    }
+
+    *S_shape->add_dim() = dim;
+  }
```

Comment:
Confirming dimension ordering here is correct. I.e. dimension at index 0 is highest dimension and dimension at index `A_shape.dim_size() - 1` is the lowest dimension.